### PR TITLE
Feature/2016 refactor kudos

### PR DESCRIPTION
**Pull request by weiqigao**
Brought the old feature/2016-refactoring-kudos branch up-to-date with the current develop branch.

### CODE REVIEW

**Comment by timyates** [comment]
I think the branch has to be called `feature-2016/refactor-kudos` to run the tests in the Github Action

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/Kudos.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+@Entity
+@Introspected
+@Table(name = "kudos")
+public class Kudos {
+
+    @Id
+    @Column(name = "id")
+    @AutoPopulated
+    @TypeDef(type = DataType.STRING)
+    @Schema(description = "the id of the kudos", required = true)
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/Kudos.java:32]: The `required` attribute is deprecated and can be removed from all of these `@Schema` annotations (it's automatically calculated now, with `requiredMode` used when overriding the value is required)

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+@Controller("/services/kudos")
+@Secured(SecurityRule.IS_AUTHENTICATED)
+@ExecuteOn(TaskExecutors.BLOCKING)
+@Tag(name = "kudos")
+public class KudosController {
+
+    private final KudosServices kudosServices;
+
+    private static final Logger LOG = LoggerFactory.getLogger(KudosController.class);
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosController.java:37]: This field is unused and can be removed

There are also a lot of unused imports in this class which we should clear up

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosController.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Get("/{id}")
+    public KudosResponseDTO getById(@NotNull UUID id) {
+        return kudosServices.getById(id);
+    }
+
+    @Get("/{?recipientId,?senderId,?isPending,?Public}")
+    public List<KudosResponseDTO> get(@Nullable UUID recipientId, @Nullable UUID senderId, @Nullable Boolean isPending, @Nullable Boolean Public) {
+        return kudosServices.findByValues(recipientId, senderId, isPending, Public);
+    }
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosController.java:63]:
```suggestion
    @Get("/{?recipientId,senderId,isPending,isPublic}")
    public List<KudosResponseDTO> get(@Nullable UUID recipientId, @Nullable UUID senderId, @Nullable Boolean isPending, @Nullable Boolean isPublic) {
        return kudosServices.findByValues(recipientId, senderId, isPending, isPublic);
    }
```

Changing to `isPublic` seems to fit better with both Java and the `isPending` parameter here...  I don't believe this is used in the front-end currently (I can find a use of `isPending`, but not `isPublic`), so I think we can just change it 🤔

Comment (timyates) [reply server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosController.java:63]: Actually, I don't think the `Public` parameter is used in the `KudosServicesImpl.findByValues` method, so maybe we can just remove it from the api 🤔

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        // Delete all KudosRecipients associated with this kudos
+        List<KudosRecipient> recipients = kudosRecipientServices.getAllByKudosId(kudos.getId());
+        kudosRecipientRepository.deleteAll(recipients);
+
+        kudosRepository.deleteById(id);
+        return true;
+    }
+
+    @Override
+    public List<KudosResponseDTO> findByValues(@Nullable UUID recipientId, @Nullable UUID senderId, @Nullable Boolean isPending, @Nullable Boolean Public) {
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:173]: I don't think `Public` is used, can we just remove it?

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        if (isPending) {
+            kudosList.addAll(kudosRepository.getAllPending());
+        } else {
+            kudosList.addAll(kudosRepository.getAllApproved());
+        }
+
+        return kudosList
+                .stream()
+                .map(this::constructKudosResponseDTO)
+                .collect(Collectors.toList());
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:214]:
```suggestion
                .toList();
```

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        List<KudosRecipient> kudosRecipients = kudosRecipientRepository.findByMemberId(memberId);
+
+        List<KudosResponseDTO> kudosList = new ArrayList<>();
+
+        kudosRecipients.forEach(kudosRecipient -> {
+            Kudos relatedKudos = kudosRepository.findById(kudosRecipient.getKudosId()).orElseThrow(() -> {
+                throw new NotFoundException("Kudos with id %s does not exist");
+            });
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:234]:
```suggestion
            Kudos relatedKudos = kudosRepository.findById(kudosRecipient.getKudosId()).orElseThrow(() ->
                new NotFoundException("Kudos with id %s does not exist")
            );
```

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            throw new PermissionException("You are not authorized to retrieve the kudos another user has sent");
+        }
+
+        String senderIdString = Util.nullSafeUUIDToString(senderId);
+        List<Kudos> kudosList = kudosRepository.search(senderIdString, true);
+
+        return kudosList
+                .stream()
+                .map(this::constructKudosResponseDTO)
+                .collect(Collectors.toList());
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:259]:
```suggestion
                .toList();
```

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        List<KudosRecipient> recipients = kudosRecipientServices.getAllByKudosId(kudos.getId());
+
+        if (recipients.isEmpty()) {
+            throw new NotFoundException("Could not find recipients for kudos with id %s");
+        }
+
+        if (kudos.getTeamId() != null) {
+            Team recipientTeam = teamRepository.findById(kudos.getTeamId()).orElseThrow(() -> {
+                throw new NotFoundException("Team %s does not exist");
+            });
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:281]:
```suggestion
            Team recipientTeam = teamRepository.findById(kudos.getTeamId()).orElseThrow(() ->
                new NotFoundException("Team %s does not exist")
            );
```

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        this.memberProfileRetrievalServices = memberProfileRetrievalServices;
+        this.currentUserServices = currentUserServices;
+    }
+
+    @Override
+    public Kudos save(KudosCreateDTO kudosDTO) {
+
+        memberProfileRetrievalServices.getById(kudosDTO.getSenderId()).orElseThrow(() -> {
+            throw new BadArgException("Kudos sender %s does not exist");
+        });
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:58]:
```suggestion
        memberProfileRetrievalServices.getById(kudosDTO.getSenderId()).orElseThrow(() ->
            new BadArgException("Kudos sender %s does not exist")
        );
```

`orElseThrow` throws the exception returned by the lambda, the lambda does not itself throw the exception (although the end result is probably the same) 😉

File: server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+        if (kudosDTO.getTeamId() != null) {
+            teamRepository.findById(kudosDTO.getTeamId()).orElseThrow(() -> {
+               throw new BadArgException("Team %s does not exist");
+            });
+        }
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java:64]:
```suggestion
        if (kudosDTO.getTeamId() != null) {
            teamRepository.findById(kudosDTO.getTeamId()).orElseThrow(() ->
               new BadArgException("Team %s does not exist")
            );
        }
```

**Comment by timyates** [comment]
Server side tests are missing are these in progress?

**Comment by mkimberlin** [comment]
@weiqigao oops...didn't mean to close this PR. It happened as a result of renaming the branch.  Do the following, and I'll open a PR for the renamed branch:

```
 git branch -m feature/2016-refactor-kudos feature-2016/refactor-kudos
 git fetch origin
 git branch -u origin/feature-2016/refactor-kudos feature-2016/refactor-kudos
 git remote set-head origin -a
 ```

**Pull request closed without merging**